agrirouter-sdk-dotnet-standard-impl/Service/Messaging/Abstraction/SendMessageBaseService.cs

```diff
@@ -86,6 +87,83 @@ public MessagingResult Send(SendMessageParameters sendMessageParameters)
+        /// </summary>
+        /// <param name="sendMessageParameters">-</param>
+        /// <returns>-</returns>
+        public Task<MessagingResult> SendAsync(SendMessageParameters sendMessageParameters)
+        {
+            var encodedMessages = new List<string>();
+
+            if (string.IsNullOrWhiteSpace(sendMessageParameters.Base64MessageContent))
+            {
+                throw new CouldNotSendEmptyMessageException("Sending empty messages does not make any sense.");
+            }
+
+            if (MessageCanBeChunked(sendMessageParameters.TechnicalMessageType))
+            {
+                if (MessageHasToBeChunked(sendMessageParameters.Base64MessageContent))
+                {
+                    var chunkContextId = Guid.NewGuid().ToString();
+                    var totalSize = Encoding.Unicode.GetByteCount(sendMessageParameters.Base64MessageContent);
+
+                    var chunkedMessages = ChunkMessageContent(sendMessageParameters.Base64MessageContent,
+                        sendMessageParameters.ChunkSize > 0
+                            ? sendMessageParameters.ChunkSize
+                            : ChunkSizeDefinition.MaximumSupported);
+
+                    var current = 0;
+                    foreach (var chunkedMessage in chunkedMessages)
+                    {
+                        var sendMessageParametersDuplicate = new SendChunkedMessageParameters
+                        {
+                            Recipients = sendMessageParameters.Recipients,
+                            TypeUrl = sendMessageParameters.TypeUrl,
+                            TechnicalMessageType = sendMessageParameters.TechnicalMessageType,
+                            ApplicationMessageId = MessageIdService.ApplicationMessageId()
+                        };
+                        var chunkComponent = new ChunkComponent
+                        {
+                            Current = current++,
+                            Total = chunkedMessage.Length,
+                            ContextId = chunkContextId,
+                            TotalSize = totalSize
+                        };
+                        sendMessageParametersDuplicate.ChunkInfo = chunkComponent;
+                        sendMessageParametersDuplicate.Base64MessageContent = chunkedMessage;
+                        encodedMessages.Add(Encode(sendMessageParametersDuplicate).Content);
+                    }
+                }
+                else
+                {
+                    encodedMessages = new List<string> {Encode(sendMessageParameters).Content};
+                }
+            }
+            else
+            {
+                encodedMessages = new List<string> {Encode(sendMessageParameters).Content};
+            }
+
+            var messagingParameters = sendMessageParameters.BuildMessagingParameter(encodedMessages);
+            return _messagingService.SendAsync(messagingParameters);
+        }
```

Comment:
This is a massive duplicate as well, I would prefer an extraction that could be used in both methods at this point.

```suggestion
 /// <summary>
        ///     Please see base class declaration for documentation.
        /// </summary>
        /// <param name="sendMessageParameters">-</param>
        /// <returns>-</returns>
        public Task<MessagingResult> SendAsync(SendMessageParameters sendMessageParameters)
        {
            var messagingParameters = CreateMessagingParameters(sendMessageParameters);
            return _messagingService.SendAsync(messagingParameters);
        }

        private MessagingParameters CreateMessagingParameters(SendMessageParameters sendMessageParameters)
        {
            var encodedMessages = new List<string>();

            if (string.IsNullOrWhiteSpace(sendMessageParameters.Base64MessageContent))
            {
                throw new CouldNotSendEmptyMessageException("Sending empty messages does not make any sense.");
            }

            if (MessageCanBeChunked(sendMessageParameters.TechnicalMessageType))
            {
                if (MessageHasToBeChunked(sendMessageParameters.Base64MessageContent))
                {
                    var chunkContextId = Guid.NewGuid().ToString();
                    var totalSize = Encoding.Unicode.GetByteCount(sendMessageParameters.Base64MessageContent);

                    var chunkedMessages = ChunkMessageContent(sendMessageParameters.Base64MessageContent,
                        sendMessageParameters.ChunkSize > 0
                            ? sendMessageParameters.ChunkSize
                            : ChunkSizeDefinition.MaximumSupported);

                    var current = 0;
                    foreach (var chunkedMessage in chunkedMessages)
                    {
                        var sendMessageParametersDuplicate = new SendChunkedMessageParameters
                        {
                            Recipients = sendMessageParameters.Recipients,
                            TypeUrl = sendMessageParameters.TypeUrl,
                            TechnicalMessageType = sendMessageParameters.TechnicalMessageType,
                            ApplicationMessageId = MessageIdService.ApplicationMessageId()
                        };
                        var chunkComponent = new ChunkComponent
                        {
                            Current = current++,
                            Total = chunkedMessage.Length,
                            ContextId = chunkContextId,
                            TotalSize = totalSize
                        };
                        sendMessageParametersDuplicate.ChunkInfo = chunkComponent;
                        sendMessageParametersDuplicate.Base64MessageContent = chunkedMessage;
                        encodedMessages.Add(Encode(sendMessageParametersDuplicate).Content);
                    }
                }
                else
                {
                    encodedMessages = new List<string> {Encode(sendMessageParameters).Content};
                }
            }
            else
            {
                encodedMessages = new List<string> {Encode(sendMessageParameters).Content};
            }

            var messagingParameters = sendMessageParameters.BuildMessagingParameter(encodedMessages);
            return messagingParameters;
        }
```

Reply:
Fixed